[JENKINS-40098] Bundle naming strategy should be able to specify an instance type

batmat opened the pull request:
[JENKINS-40098](https://issues.jenkins-ci.org/browse/JENKINS-40098)

Refactored the existing code to use the same logic for choosing the bundle filename, and introduced a change enabling both overriding through using a sysprop, or implementing a new Extension Point.

The approach with Extension Points should enable more dynamic needs, when the sysprop one will be immediately usable by anyone without having to develop an extension.

Note: if found, the extension is preferred over the sysprop.

@reviewbybees esp. @jglick who pointed me to SupportProvider.getName() I had missed in the first attempt. 

Supersedes #91

jtnord commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,78 @@
+package com.cloudbees.jenkins.support;
```
(c) header please

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jtnord commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,78 @@
+
+        if (all.size() > 1) {
+            if (LOGGER.isLoggable(Level.WARNING)) {
+                LOGGER.warning("{0} implementations found for support bundle prefix naming strategy. " +
```
:bug: you are missing a parameter here.
NIT: from a stylistic point I think its better to always use Logger.log(Level.XXX ...) rather than the Logger.xxx(...)

jtnord commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,78 @@
+            return DEFAULT_STRATEGY;
+        }
+
+        if (all.size() > 1) {
```
why not use `extensionCount`

jtnord commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return the instance type specification to be used for generated support bundles.
+     */
+    @Nonnull
```
is this really the annotations that this code already uses.... If not then swtich.

amuniz commented on src/test/java/com/cloudbees/jenkins/support/BundleNamePrefixTest.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Test
+    public void checkWithOneProvider() throws Exception {
+        assertThat(SupportPlugin.getBundleFileName(), startsWith("support_pouet_" + CURRENT_YEAR));
```
Oh, ok, I see the method name match with the `@TestExtension` parameter.

jtnord commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static final Logger LOGGER = Logger.getLogger(BundleNameInstanceTypeProvider.class.getName());
+
+    private static final BundleNameInstanceTypeProvider DEFAULT_STRATEGY = new BundleNameInstanceTypeProvider() {
```
why not have this as an extension with very low priority and avoid the extra complexity below?

batmat replied:
Just because I wasn't absolutely sure about the behaviour of the priority, going to check, and if that works and the doc confirms, will do.

jtnord commented on src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java:
```diff
@@ -763,6 +774,33 @@ public boolean accept(File dir, String name) {
+                    SupportPlugin.class.getSimpleName(), new Date()));
+            File[] files = bundleDir.listFiles(new FilenameFilter() {
+                public boolean accept(File dir, String name) {
+                    return name.startsWith(SupportPlugin.getBundlePrefix()) && name.endsWith(".zip");
```
this will fail to clean up old bundles if the provider is changed.

Why not just check for "support*.zip"?

batmat replied:
Because the SupportProvider.getName() could make that pattern totally different anyway, files could be named `whatever-thedate.zip`.

A different approach might be to just select every zip files like `$JENKINS_HOME/support/*.zip`, but this is a bit risky if someone has taken the habit to put unrelated files there...

jglick replied:
> if someone has taken the habit to put unrelated files there

Well don’t do that then.

I agree it would be better to just delete all `*.zip`.

batmat replied:
done

jtnord commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,78 @@
+                LOGGER.warning("{0} implementations found for support bundle prefix naming strategy. " +
+                        "Can be only 0 or 1. Choosing the first found.");
+                for (BundleNameInstanceTypeProvider bundlePrefixProvider : all) {
+                    LOGGER.log(Level.WARNING, "class '{0}' found", bundlePrefixProvider.getClass().getName());
```
:bug:  `'` needs to be escaped other wise this will log `"class {0} found`"
https://docs.oracle.com/javase/7/docs/api/index.html?java/text/MessageFormat.html

batmat commented:
@jtnord I believe I addressed all your comments, but https://github.com/jenkinsci/support-core-plugin/pull/92#discussion_r90266258. For that one, I guess the only "risk" is to have uncleaned lefovers that users may have to clean up manually. The alternative is to work more largely, but we take then IMO an even more risky approach: wiping out unrelated zip files. WDYT?

If you have a better idea or think this is fine to just consider nobody should have put things in that directory (which, well, I would agree on...) then we can make it simple(r).

amuniz commented:
:bee:

jglick commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,106 @@
+ * <ol>
+ * <li>If an implementation of {@link BundleNameInstanceTypeProvider} is found, it will be used.<br>
+ * <strong>WARNING: </strong>if many are found, then a warning will be issued, and the first extension found will
+ * be used.</li>
```
So, where is this nondefault implementation? For any newly introduced API I expect to see a PR from another plugin repository using a `SNAPSHOT` dependency which consumes that API. (Obviously public use cases are preferred whenever possible, but note that a PR in a private repository may refer in its summary etc. to a public PR, in which case GitHub automatically creates a backlink in the public PR visible to users who have permission to view the private repo.)

christ66 commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,106 @@
+    public abstract String getInstanceType();
+
+    // We want this to be always picked up last in case others are found Double.MIN_VALUE does NOT work
+    @Extension(ordinal = -1000000)
```
`@Extension(ordinal = Integer.MAX_VALUE)`

batmat replied:
Well, why not, but this is still weird, since the type is `double` anyway...

batmat replied:
fixed, though I don't really like it. But in the end with the comment above, ppl puzzled by it should understand.

jglick commented on src/main/java/com/cloudbees/jenkins/support/SupportCommand.java:
```diff
@@ -107,7 +98,7 @@ protected int run() throws Exception {
 
         public OutputStream call() throws IOException {
-            File f = File.createTempFile(filename, ".zip");
+            File f = File.createTempFile(filename, "");
```
This is wrong. Even if `filename` ends in `.zip`, you will end up with something like this:

```
$ jrunscript -e 'println(java.io.File.createTempFile("foo.zip", ""))'
/tmp/foo.zip6023275810588304403
```

jglick replied:
Though if we are always adding the current date down to second resolution anyway, then perhaps we can omit the generated suffix and just save `${java.io.tmpdir}/${filename}`.

batmat replied:
Done. I settled on using 

```java
Path path = Files.createFile(Paths.get(System.getProperty("java.io.tmpDir"), filename));
```

jglick commented on src/main/java/com/cloudbees/jenkins/support/SupportCommand.java:
```diff
@@ -107,7 +98,7 @@ protected int run() throws Exception {
+            File f = File.createTempFile(filename, "");
             System.err.println("Creating: " + f);
             return new RemoteOutputStream(new FileOutputStream(f));
         }
```
BTW not introduced in this PR, but while you are here please delete the bogus `checkRoles` implementation and extend `MasterToSlaveCallable`.

batmat replied:
done

jglick commented on src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java:
```diff
@@ -563,6 +561,46 @@ public void run() {
+        dateFormat.setTimeZone(TimeZone.getTimeZone("GMT"));
+        filename.append("_").append(dateFormat.format(new Date()));
+
+        filename.append(".zip");
```
As noted above, better to have the extension appended by the caller.

batmat replied:
so, finally used `filename` directly instead of a temp file created by the dedicated API.

jglick commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * Returns the <strong>non-null</strong> instance type to be used for generated support bundle names.
```
You need to document that the empty string is allowed, and in fact the default.

batmat replied:
fixed. I actually explained it's expected to be not empty (contrary to the default value)

jglick commented on src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java:
```diff
@@ -563,6 +561,46 @@ public void run() {
+            SupportProvider supportProvider = instance.getSupportProvider();
+            if (supportProvider != null) {
+                // let the provider name it
+                filename = supportProvider.getName();
```
Still leaves me wondering about why we have both systems. In the case of a vendor Acme implementing both extension points in a Blaster product, we will get something like

    acme-support_acme-blaster_2016-12-07_16.12.47.zip

which seems redundant.

batmat replied:
Well, in that case, that vendor would just want to not put Acme everywhere. Which would be more:

`acme-support_blaster_2016-12-07_16.12.47.zip`

which seems less redundant than your example?

batmat commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,106 @@
+
+        final BundleNameInstanceTypeProvider chosen = all.get(0);
+        if (extensionCount > 1) {
+            LOGGER.log(Level.INFO, "Using {0} as BundleNameInstanceTypeProvider implementation", chosen.getClass().getName());
```
fixed

batmat commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,106 @@
+     * Aims to provide informational data about the generated bundles.
+     * <p>
+     * <p>
+     * <p><b>Will be used for file name generation, so avoid funky characters.
```
fixed

batmat commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,106 @@
+     * <p>
+     * <p>
+     * <p><b>Will be used for file name generation, so avoid funky characters.
+     * Please stay in <code>[a-zA-Z-_]</code></b>. Also consider the file name length, you probably want to be defensive
```
fixed

christ66 commented on src/main/java/com/cloudbees/jenkins/support/SupportCommand.java:
```diff
@@ -99,23 +91,18 @@ protected int run() throws Exception {
-            // TODO: do we have to verify some role?
+        public OutputStream call() throws IOException {
+            Path path = Files.createFile(Paths.get(System.getProperty("java.io.tmpDir"), filename));
+            System.err.println("Creating: " + path);
```
🐛 `System.err`

christ66 replied:
Ah wait this is for the SupportCommand, nvm

jglick commented on src/main/java/com/cloudbees/jenkins/support/SupportCommand.java:
```diff
@@ -90,23 +91,18 @@ protected int run() throws Exception {
-        public void checkRoles(RoleChecker checker) throws SecurityException {
-            // TODO: do we have to verify some role?
+        public OutputStream call() throws IOException {
+            Path path = Files.createFile(Paths.get(System.getProperty("java.io.tmpDir"), filename));
```
It is `tmpdir` not `tmpDir`. Did you try running this command?

jglick commented on src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java:
```diff
@@ -0,0 +1,108 @@
+        final int extensionCount = all.size();
+
+        if (extensionCount > 2) {
+            if (LOGGER.isLoggable(Level.WARNING)) {
```
We log `INFO` by default. So no, it is not terribly useful, but harmless I suppose.

batmat commented:
Fixed @jglick yes I tried it locally using jrunscript, but didn't adjusted after fixing the jrunscript. Sorry for the noise.

jglick commented:
> I tried it locally using jrunscript

Not sure what you mean. You need to

    mvn hpi:run &
    wget -O /tmp/cli.jar http://localhost:8080/jenkins/jnlpJars/jenkins-cli.jar
    java -jar /tmp/cli.jar -s http://localhost:8080/jenkins/ support

jglick commented on src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java:
```diff
@@ -763,6 +770,38 @@ public boolean accept(File dir, String name) {
+            long pivot = System.currentTimeMillis();
+            for (long l = 1; l * 2 > 0; l *= 2) {
+                boolean seen = false;
+                for (File f : files) {
```
NPE if `listFiles` returns null, as it is permitted to do even if you believe it currently exists.

batmat replied:
@jglick I just moved that code in a dedicated method to clarify the code, didn't touch it. But will happily file a separate PR about it.
Thanks for the review !

christ66 commented:
@batmat @jglick @jtnord I am wondering if this PR should be reverted. This only makes the naming of the support bundles significantly longer.

Before this PR my support bundles use to be:

```
cloudbees-support_2017-02-02_18.16.45
```

which is already long enough and unique (as in the time it was created IMO is random and unique enough), but now they are:

```
cloudbees-support_UNKNOWN-localhost_8080_jenkins_2017-01-24_20.02.19
```

Having this long of a name for a folder I do not believe is right. Just a lot of extra noise which does not need to exist in the name.

jtnord commented:
not sure why that comes up as `UNKNOWN` but the name is really really really useful when you have lots of bundles from the entire cluster in a support ticket that are updated at various occasions during the lifecycle...
The fact it says` cloudbees-support` would possibly imply you are running a cloudbees distribution without some required plugins that should be installed by default on every instance?
